[cli/trampolines] Clean up definitions, fix win32 exporting issue

staticfloat opened the pull request:
Our win32 trampolines weren't getting exported properly.  Also take the
opportunity to properly organize after debugging libblastrampoline and
finding the current organizational strategy lacking.

In the future, if we ever want to mangle the symbols from these
trampolines, we now have a convenient `CNAME()` that is properly invoked
from all trampolines and can consistently mangle names.

staticfloat committed:
[cli/trampolines] Clean up definitions, fix win32 exporting issue

Our win32 trampolines weren't getting exported properly.  Also take the
opportunity to properly organize after debugging libblastrampoline and
finding the current organizational strategy lacking.

In the future, if we ever want to mangle the symbols from these
trampolines, we now have a convenient `CNAME()` that is properly invoked
from all trampolines and can consistently mangle names.

vtjnash commented on cli/trampolines/common.h:
```diff
@@ -0,0 +1,70 @@
+
+// aarch64 on mac requires some special assembler syntax for both calculating memory
+// offsets and even just the assembler statement separator token
+#if defined(__aarch64__)
```
why wouldn't this go in trampolines_aarch64.S, when it is only defined and used there?

staticfloat replied:
because it's nice to have the `.S` files be as free of preprocessor stuff as possible; it makes it much easier to, with your eyeball, compare the trampolines across platforms.

staticfloat committed:
[cli/trampolines]: Don't clobber `r9` on powerpc64le

According to the examples on page 138-139 of the ELV v2 OpenPOWER ABI [0],
we should avoid clobbering r9 and only touch r2 and r12.  This fixed a
test failure in libblastrampoline, so it's probably needed here as well.

[0]: https://members.openpowerfoundation.org/document/dl/576

staticfloat commented:
Adding @vchuravy because I changed the ppc64le trampoline again; it clobbers one less register now and passes libblastrampoline tests on power now, so I figure we probably want that change here as well.

vchuravy reviewed (approved):
The PPC changes look right and necessary. `r9` is volatile and used to pass arguments.

staticfloat committed:
Fix `armv7l` trampoline
